src/android/com/mady/wifi/api/WifiStatus.java

```diff
@@ -284,13 +287,39 @@ public boolean pingCmd(String addr) {
             int exit = pro.exitValue();
             if (exit == 0) {
                 return true;
-            } else {
-                //ip address is not reachable
-                return false;
             }
         } catch (IOException e) {
+            Log.d("RAVEN", "ICMP ping error - " + e.getMessage());
+        }
+        Log.d("RAVEN", "ICMP ping to " + addr + " failed");
+        if (isReachable(addr, 443, 5000)) return true;
+        return isReachable(addr, 80, 5000);
+    }
+
+    /**
+     * Checks via TCP connection if we are connected to the internet. More reliable than using native Android 
+     * InetAddress.isReachable functionality because we are allowed to specify port (rather than only using port 7).
+     * Taken from - https://stackoverflow.com/a/34228756/13173970
+     * 
+     * @param addr IP address you want to check
+     * @param openPort port to check for connection
+     * @param timeOutMillis timeout in ms
+     * @return true if the IP address is reachable, false if not
+     */
+    private boolean isReachable(String addr, int openPort, int timeOutMillis) {
+        // Any Open port on other machine
+        // openPort =  22 - ssh, 80 or 443 - webserver, 25 - mailserver etc.
+        try {
+            Log.d("RAVEN", "Trying TCP connection to " + addr + " over port " + String.valueOf(openPort));
+            try (Socket soc = new Socket()) {
+                // throws IOException if connection cannot be established on port
+                soc.connect(new InetSocketAddress(addr, openPort), timeOutMillis);
```

Comment:
There are two other exceptions that could potentially bubble up from here:

    IllegalBlockingModeException - if this socket has an associated channel, and the channel is in non-blocking mode
    IllegalArgumentException - if endpoint is null or is a SocketAddress subclass not supported by this socket

Even though they "shouldn't" occur, it is a good idea to handle them appropriately just in case they ever do.